src/connector/src/sink/big_query.rs

```diff
@@ -678,49 +645,74 @@ impl StorageWriterClient {
+        let stream = tokio_stream::wrappers::ReceiverStream::new(rx);
+        let (result_sender, _) = broadcast::channel(BIGQUERY_SEND_FUTURE_BUFFER_MAX_SIZE + 1);
+        let result_sender = Arc::new(result_sender);
+        let result_sender_clone = result_sender.clone();
+        let abort_handle = tokio::spawn(async move {
+            let mut resp = client
+                .append_rows(Request::new(stream))
+                .await
+                .map_err(|e| SinkError::BigQuery(e.into()))
+                .unwrap()
+                .into_inner();
+            let mut offset = 0;
+            loop {
+                let result = match resp.message().await {
+                    Ok(Some(append_rows_response)) => {
+                        if !append_rows_response.row_errors.is_empty() {
+                            Some(format!(
+                                "Insert error {:?}",
+                                append_rows_response.row_errors
+                            ))
+                        } else {
+                            None
+                        }
+                    }
+                    Ok(None) => {
+                        continue;
+                    }
+                    Err(e) => Some(e.to_report_string()),
+                };
+                result_sender_clone.send((result, offset)).unwrap();
+                offset += 1;
+            }
+        })
+        .abort_handle();
+
         Ok(StorageWriterClient {
-            client,
             environment,
+            request_sender: tx,
+            result_sender,
+            abort_handle,
+            offset: 0,
         })
     }
 
+    pub fn get_subscribe(&mut self) -> (u64, broadcast::Receiver<(Option<String>, u64)>) {
+        self.offset += 1;
+        (self.offset - 1, self.result_sender.subscribe())
+    }
+
     pub async fn append_rows(
         &mut self,
-        rows: Vec<AppendRowsRequestRows>,
+        row: AppendRowsRequestRows,
         write_stream: String,
     ) -> Result<()> {
-        let mut resp_count = rows.len();
-        let append_req: Vec<AppendRowsRequest> = rows
-            .into_iter()
-            .map(|row| AppendRowsRequest {
-                write_stream: write_stream.clone(),
-                offset: None,
-                trace_id: Uuid::new_v4().hyphenated().to_string(),
-                missing_value_interpretations: HashMap::default(),
-                rows: Some(row),
-            })
-            .collect();
-        let mut resp = self
-            .client
-            .append_rows(Request::new(tokio_stream::iter(append_req)))
-            .await
-            .map_err(|e| SinkError::BigQuery(e.into()))?
-            .into_inner();
-        while let Some(append_rows_response) = resp
-            .message()
+        let append_req = AppendRowsRequest {
+            write_stream: write_stream.clone(),
+            offset: None,
+            trace_id: Uuid::new_v4().hyphenated().to_string(),
```

Comment:
Can you tracked the `trace_id` locally instead of incrementing the `self.offset`? I think in the response stream, it should return the same trace_id in the corresponding response.

Reply:
Tried it, but there is no ID in the returned message